File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/EnumTranslatorUtil.java
package com.cedarsoft.commons.javafx;

import java.util.concurrent.atomic.AtomicReference;

import javax.annotation.Nonnull;

/**
 * @author Christian Erbelding (<a href="mailto:ce@cedarsoft.com">ce@cedarsoft.com</a>)
 */
public class EnumTranslatorUtil {
  /**
   * Holds the reference to the currently active translator that is set in a static field
   */
  @Nonnull
  private static final AtomicReference<EnumTranslator> ENUM_TRANSLATOR = new AtomicReference<>(item -> {
    System.err.println("No enum translator set in com.cedarsoft.commons.javafx.EnumListCell");
    return item.name();
  });

  private EnumTranslatorUtil() {
    // utility class
  }

  @Nonnull
  public static EnumTranslator getEnumTranslator() {
    return ENUM_TRANSLATOR.get();
  }

  public static void setEnumTranslator(@Nonnull EnumTranslator enumTranslator) {
    ENUM_TRANSLATOR.set(enumTranslator);
  }

}

File: commons/javafx/src/main/kotlin/com/cedarsoft/commons/javafx/EnumTranslatorUtil.kt
package com.cedarsoft.commons.javafx

import it.neckar.open.annotations.JavaFriendly
import javax.annotation.Nonnull

typealias EnumTranslatorLambda = (item: Enum<*>) -> String?

/**
 * Enum translator
 */
object EnumTranslatorUtil {
  /**
   * Fallback enum translator
   */
  private val fallback: EnumTranslatorLambda = { item ->
    System.err.println("No enum translation found for <$item>")
    item.name
  }

  /**
   * The enum translators. The first one that returns a result will be used
   */
  @Nonnull
  private val enumTranslatorsChain = mutableListOf(fallback)

  /**
   * Adds an enum translator at *first* position.
   * The given enum translator is used *first*
   */
  @JvmStatic
  @JavaFriendly
  fun addEnumTranslator(enumTranslator: EnumTranslator) {
    enumTranslator(enumTranslator::translate)
  }

  /**
   * Adds an enum translator
   */
  fun enumTranslator(enumTranslator: EnumTranslatorLambda) {
    enumTranslatorsChain.add(0, enumTranslator)
  }

  /**
   * Resets the chain. Just adds the fallback enum translator
   */
  @JvmStatic
  fun reset() {
    enumTranslatorsChain.clear()
    enumTranslatorsChain.add(fallback)
  }

  /**
   * Translates the num
   */
  @JvmStatic
  fun translate(item: Enum<*>): String {
    return enumTranslatorsChain.firstNotNullOf { it(item) }
  }

  /**
   * The enum translator that uses the translator chain
   */
  @JvmStatic
  val enumTranslator: EnumTranslator = object : EnumTranslator {
    override fun translate(item: Enum<*>): String {
      return EnumTranslatorUtil.translate(item)
    }
  }
}
